llvm/lib/Target/X86/X86ISelLowering.cpp

```
@@ -7230,6 +7234,20 @@ static bool findEltLoadSrc(SDValue Elt, LoadSDNode *&Ld, int64_t &ByteOffset) {
       }
     }
     break;
+  case ISD::EXTRACT_ELEMENT:
+    if (auto *IdxC = dyn_cast<ConstantSDNode>(Elt.getOperand(1))) {
```

Pretty sure this must be a constant. But this could also be done separately, it's not related to the atomic 

But also, can we avoid this by not using EXTRACT_ELEMENT in the first place? EXTRACT_ELEMENT has weird handling I've never understood where it's hardly used

---

If it isn't a constant, shall we assert false or abort the transform? Right now it aborts the transform. If we know it always is and will be constant, I guess asserting is better. Separately as in another PR? There would be no associated test change as this is required for the optimization.

Yes, if we use the ones already implemented here, then we will be able to discover the `ByteOffset`. Why do we want to avoid EXTRACT_ELEMENT? It seems to work here.

---

> If it isn't a constant, shall we assert false or abort the transform?

Ignore it as a possibility and just let it crash. No efforts should be made to support invalid constructs 

> Why do we want to avoid EXTRACT_ELEMENT? It seems to work here.

Of course it works, but seems poorly supported by optimizations. In particular the comment on it says "This is only for use before legalization, for values that will be broken into multiple registers.". This is use during legalization (dont' really know why we have this, or this fake restriction documented). You could get the same with just shifts.